File: src/distrib.cpp
// distrib.cpp

#include "bayesmr.h"

// ------------------------------------------------------------
// Internal helpers (not visible outside this .cpp)
// ------------------------------------------------------------
namespace {

std::size_t recycle_size(std::size_t a,
                         std::size_t b,
                         std::size_t c = 1,
                         std::size_t d = 1){
  return std::max({a, b, c, d});
}

double recycle_get(const std::vector<double>& v, std::size_t i){
  return v[i % v.size()];
}

} // unnamed namespace

// Probability function for the product of independent bernoulli random variables
void dprodber(double* prob, const int* d, const double* pi, int m, bool logscale = true){
  double prob_tmp = 0;
  if(!logscale){
    *prob = 0;
    for(int i = 0; i < m; i++){
      prob_tmp = d[i]*log(pi[i]) + (1 - d[i])*log(1 - pi[i]);
      if(ISNAN(prob_tmp)){
        prob_tmp = log(pow(pi[i], d[i])*pow(1 - pi[i], 1 - d[i]));
      }
      *prob += prob_tmp;
    }
  }
  else{
    *prob = 1;  
    for(int i = 0; i < m; i++){
      prob_tmp = pow(pi[i], d[i])*pow(1 - pi[i], 1 - d[i]);
      *prob *= prob_tmp;
    }
  }
}

// Density function for a multivariate normal random variable
arma::vec dmvnorm_fast(const arma::mat& X,
                       const arma::vec& mu,
                       const arma::mat& Sigma,
                       bool logscale = true){
  
  int n = X.n_rows;
  int k = X.n_cols;
  
  // Pre-compute Cholesky and constants
  arma::mat L = arma::chol(Sigma, "lower");
  double log_det_sigma = 2.0 * arma::sum(arma::log(L.diag()));
  
  arma::vec result(n);
  
  for(int i = 0; i < n; i++){
    arma::vec diff = X.row(i).t() - mu;
    arma::vec z = arma::solve(arma::trimatl(L), diff);
    
    result(i) = -0.5 * (k * log_2pi + log_det_sigma + arma::dot(z, z));
  }
  
  if(!logscale){
    result = arma::exp(result);
  }
  
  return result;
}

// Density function for a multivariate normal random variable
void dmultinorm(double* log_dens, const double* x, const double* mean, const double* sigma, int n, int p,
                bool logscale = true){
  arma::mat x_arma(n, p);
  arma::vec mean_arma(p);
  arma::mat sigma_arma(p, p);
  for(int j = 0; j < p; j++){
    for(int i = 0; i < n; i++){
      x_arma(i, j) = x[n*j + i];
    }
    mean_arma(j) = mean[j];
    for(int k = 0; k < p; k++){
      sigma_arma(j, k) = sigma[p*k + j];
    }
  }

  double logdet, sign;
  arma::log_det(logdet, sign, sigma_arma);
  arma::vec distval = mahalanobis(x_arma, mean_arma, sigma_arma);
  for(int i = 0; i < n; i++){
    log_dens[i] = -0.5*(p * log(2.0*M_PI) + logdet + distval[i]);
  }

  if(!logscale){
    for(int i = 0; i < n; i++){
      log_dens[i] = exp(log_dens[i]);
    }
  }
}

// Multivariate normal random deviates
void rmultinorm(double* dev, int n, const double* mean, const double* sigma, int p){
  arma::vec mean_arma(p);
  arma::mat sigma_arma(p, p);
  for(int j = 0; j < p; j++){
    mean_arma(j) = mean[j];
    for(int i = 0; i < p; i++){
      sigma_arma(i, j) = sigma[i + p*j];
    }
  }

  arma::mat V(p, p);
  arma::mat U(p, p);
  arma::vec d(p);
  arma::svd(U, d, V, sigma_arma);
  arma::mat U_trans = trans(U);

  arma::mat tmp(p, p);
  for(int i = 0; i < p; i++){
    for(int j = 0; j < p; j++){
      tmp(i, j) = U_trans(i, j)*sqrt(d(i));
    }
  }
  arma::mat sigmaSVDsqrt = trans(V * tmp);
  arma::mat normdev(n, p);
  for(int j = 0; j < p; j++){
    for(int i = 0; i < n; i++){
      normdev(i, j) = R::rnorm(0, 1);
    }
  }
  arma::mat out_tmp = normdev * sigmaSVDsqrt;
  for(int j = 0; j < p; j++){
    for(int i = 0; i < n; i++){
      dev[n*j + i] = out_tmp(i, j) + mean_arma(j);
    }
  }
}

// Density function for an inverse gamma random variable
void dinvgamma(double* log_dens, const double* x, const double alpha, const double beta, int n,
               bool logscale = true){
  if((alpha <= 0) || (beta <= 0)){
    error("alpha (shape) and beta (scale) parameters in dinvgamma() need to be both strictly positive.\n");
  }

  double lbeta = log(beta);
  double lgalpha = R::lgammafn(alpha);
  for(int i = 0; i < n; i++){
    log_dens[i] = alpha * lbeta - lgalpha - (alpha + 1) * log(x[i]) - (beta/x[i]);
  }

  if(!logscale){
    for(int i = 0; i < n; i++){
      log_dens[i] = exp(log_dens[i]);
    }
  }
}

// Inverse gamma random deviates
void rinvgamma(double* dev, int n, const double alpha, const double beta){
  if((alpha <= 0) || (beta <= 0)){
    error("alpha (shape) and beta (scale) parameters in rinvgamma() need to be both strictly positive.\n");
  }

  for(int i = 0; i < n; i++){
    dev[i] = 1.0/R::rgamma(alpha, 1.0/beta);
  }
}

// Density function for a Dirichlet random variable
void ddirichlet(double* log_dens, const double* x, const double* par, int n, int p,
                bool logscale = true){
  double tmp = 0;
  arma::mat x_arma(n, p);
  for(int j = 0; j < p; j++){
    for(int i = 0; i < n; i++){
      x_arma(i, j) = x[n*j + i];
    }
  }

  tmp = 0;
  for(int i = 0; i < n; i++){
    for(int j = 0; j < p; j++){
      tmp += x_arma(i, j);
      if(x_arma(i, j) < 0 || x_arma(i, j) > 1){
        error("some elements of x outside the [0, 1] range.\n");
      }
    }
    if(std::fabs(tmp - 1.0) >= sqrt(std::numeric_limits<double>::epsilon())){
      error("some rows of x sum to a value different from 1.\n");
    }
  }

  double logD = 0, salpha = 0, s = 0;
  for(int j = 0; j < p; j++){
    logD += R::lgammafn(par[j]);
    salpha += par[j];
  }
  logD = logD - R::lgammafn(salpha);
  for(int i = 0; i < n; i++){
    s = 0;
    for(int j = 0; j < p; j++){
      s += (par[j] - 1)*log(x_arma(i, j));
    }
    log_dens[i] = s - logD;
    if(!logscale){
      log_dens[i] = exp(log_dens[i]);
    }
  }
}

// Dirichlet random deviates
void rdirichlet(double* dev, int n, const double* par, int p){
  arma::mat z(n, p, arma::fill::zeros);
  arma::vec s(n, arma::fill::zeros);

  for(int j = 0; j < p; j++){
    for(int i = 0; i < n; i++){
      z(i, j) = R::rgamma(par[j], 1.0);
      s(i) += z(i, j);
    }
  }
  for(int i = 0; i < n; i++){
    for(int j = 0; j < p; j++){
      dev[n*j + i] = z(i, j)/s(i);
    }
  }
}

// Bivariate normal
std::vector<double> dbivnorm_cpp(const std::vector<double>& x_vec, const std::vector<double>& y_vec,
  const std::vector<double>& mu_x, const std::vector<double>& mu_y,
  const std::vector<double>& sigma_xx, const std::vector<double>& sigma_yy,
  const std::vector<double>& sigma_xy, bool logscale){
  const size_t n = x_vec.size();
  if (y_vec.size() != n || mu_x.size() != n || mu_y.size() != n ||
    sigma_xx.size() != n || sigma_yy.size() != n || sigma_xy.size() != n){
      Rcpp::stop("all input vectors must have the same length.");
  }

  std::vector<double> out(n);

  for (size_t i = 0; i < n; ++i){
    const double sxx = sigma_xx[i];
    const double syy = sigma_yy[i];
    const double sxy = sigma_xy[i];
    const double det = sxx * syy - sxy * sxy;

    // Invalid covariance?
    if (sxx <= 0.0 || syy <= 0.0 || det <= 0.0) {
      out[i] = logscale ?
        -std::numeric_limits<double>::infinity() :
        0.0;
      continue;
    }

    const double inv_det = 1.0 / det;
    const double dx = x_vec[i] - mu_x[i];
    const double dy = y_vec[i] - mu_y[i];

    const double q =
      inv_det * (syy * dx * dx - 2.0 * sxy * dx * dy + sxx * dy * dy);

    const double logdens =
      -0.5 * (log_2pi + std::log(det) + q);

    out[i] = logscale ? logdens : std::exp(logdens);
  }

  return out;
}

std::vector<double> dhalft(const std::vector<double>& x, const std::vector<double>& alpha,
  const std::vector<double>& nu, bool logscale){
  std::size_t N = recycle_size(x.size(), alpha.size(), nu.size());

  std::vector<double> out(N);

  for (double a : alpha)
    if (a <= 0.0)
      throw std::domain_error("The alpha (scale) parameter must be positive.");

  for (std::size_t i = 0; i < N; ++i) {
    double xi = recycle_get(x, i);
    double ai = recycle_get(alpha, i);
    double ni = recycle_get(nu, i);

    double log_dens =
        std::log(2.0)
        - std::log(ai)
        + R::lgammafn((ni + 1.0) / 2.0)
        - R::lgammafn(ni / 2.0)
        - 0.5 * std::log(M_PI * ni)
        - (ni + 1.0) / 2.0 * std::log(1.0 + (xi / ai) * (xi / ai) / ni);

    out[i] = logscale ? log_dens : std::exp(log_dens);
  }

  return out;
}

double dhalft_scalar(const double& x, const double& alpha,
  const double& nu, bool logscale){
  if (alpha <= 0.0)
    throw std::domain_error("The alpha (scale) parameter must be positive.");
  if (nu <= 0.0)
    throw std::domain_error("The nu (df) parameter must be positive.");

  // support
  if (x < 0.0)
    return logscale ? R_NegInf : 0.0;

  const double log_dens =
      std::log(2.0)
      - std::log(alpha)
      + R::lgammafn((nu + 1.0) / 2.0)
      - R::lgammafn(nu / 2.0)
      - 0.5 * std::log(M_PI * nu)
      - (nu + 1.0) / 2.0 *
        std::log(1.0 + (x / alpha) * (x / alpha) / nu);

  return logscale ? log_dens : std::exp(log_dens);
}

std::vector<double> pst(const std::vector<double>& q, const std::vector<double>& mu,
  const std::vector<double>& sigma, const std::vector<double>& nu,
  bool lower_tail, bool log_p){
  std::size_t N = recycle_size(q.size(), mu.size(), sigma.size(), nu.size());
  std::vector<double> out(N);

  for (double s : sigma)
    if (s <= 0.0)
      throw std::domain_error("The sigma parameter must be positive.");

  for (double n : nu)
    if (n <= 0.0)
      throw std::domain_error("The nu parameter must be positive.");

  for (std::size_t i = 0; i < N; ++i) {
    double qi = recycle_get(q, i);
    double mi = recycle_get(mu, i);
    double si = recycle_get(sigma, i);
    double ni = recycle_get(nu, i);

    if (ni > 1e6) {
      out[i] = R::pnorm(qi, mi, si, lower_tail, log_p);
    } else {
      double z = (qi - mi) / si;
      out[i] = R::pt(z, ni, lower_tail, log_p);
    }
  }

  return out;
}

std::vector<double> qst(const std::vector<double>& p, const std::vector<double>& mu,
  const std::vector<double>& sigma, const std::vector<double>& nu, bool lower_tail,
  bool log_p){
  std::size_t N = recycle_size(p.size(), mu.size(), sigma.size(), nu.size());
  std::vector<double> out(N);

  for (double s : sigma)
    if (s <= 0.0)
      throw std::domain_error("The sigma parameter must be positive.");

  for (double n : nu)
    if (n <= 0.0)
      throw std::domain_error("The nu parameter must be positive.");

  for (std::size_t i = 0; i < N; ++i) {
    double pi = recycle_get(p, i);
    double mi = recycle_get(mu, i);
    double si = recycle_get(sigma, i);
    double ni = recycle_get(nu, i);

    if (log_p)
      pi = std::exp(pi);

    if (pi < 0.0 || pi > 1.0)
      throw std::domain_error("p must be in [0,1].");

    if (ni > 1e6) {
      out[i] = R::qnorm(pi, mi, si, lower_tail, false);
    } else {
      out[i] = mi + si * R::qt(pi, ni, lower_tail, false);
    }
  }

  return out;
}

std::vector<double> qtrunc(const std::vector<double>& p, double a, double b,
  const std::vector<double>& mu, const std::vector<double>& sigma,
  const std::vector<double>& nu){
  if (a >= b)
    throw std::domain_error("Lower bound a is not less than upper bound b.");

  std::size_t N = recycle_size(p.size(), mu.size(), sigma.size(), nu.size());
  std::vector<double> out(N);

  for (std::size_t i = 0; i < N; ++i) {
    double pi = recycle_get(p, i);
    double mi = recycle_get(mu, i);
    double si = recycle_get(sigma, i);
    double ni = recycle_get(nu, i);

    if (pi < 0.0 || pi > 1.0)
      throw std::domain_error("p must be in [0,1].");

    double Ga = R::pt((a - mi) / si, ni, true, false);
    double Gb = R::pt((b - mi) / si, ni, true, false);

    out[i] = mi + si * R::qt(Ga + pi * (Gb - Ga), ni, true, false);
  }

  return out;
}

std::vector<double> rtrunc(std::size_t n, double a, double b,
  const std::vector<double>& mu, const std::vector<double>& sigma,
  const std::vector<double>& nu){
  if (a >= b)
    throw std::domain_error("Lower bound a is not less than upper bound b.");

  std::vector<double> u(n);
  for (std::size_t i = 0; i < n; ++i)
    u[i] = unif_rand();

  return qtrunc(u, a, b, mu, sigma, nu);
}

std::vector<double> rhalft(std::size_t n, const std::vector<double>& alpha, const std::vector<double>& nu){
  for (double a : alpha)
    if (a <= 0.0)
      throw std::domain_error("The alpha parameter must be positive.");

  std::vector<double> mu(1, 0.0);

  return rtrunc(n, 0.0, R_PosInf, mu, alpha, nu);
}

File: R/distrib.R
dinvgamma <- function(x, alpha, beta = 1, log = FALSE) {
	if ((alpha <= 0) | (beta <= 0)) {
		stop("alpha (shape) and/or beta (scale) parameter negative in dinvgamma().\n")
	}
	log_density <- alpha*log(beta) - lgamma(alpha) - (alpha + 1)*log(x) - (beta/x)
	if (log) return(log_density) else return(exp(log_density))
}

ddirichlet <- function(x, alpha) {
	dirichlet1 <- function(x, alpha) {
		logD <- sum(lgamma(alpha)) - lgamma(sum(alpha))
		s <- sum((alpha - 1)*log(x))
		exp(sum(s) - logD)
	}
	if (!is.matrix(x)) 
		if (is.data.frame(x))
			x <- as.matrix(x)
		else x <- t(x)
	if (!is.matrix(alpha)) 
		alpha <- matrix(alpha, ncol = length(alpha), nrow = nrow(x), byrow = TRUE)

	if (any(alpha <= 0))
		stop("the elements of the alpha vector must be strictly positive.")
	if (any(dim(x) != dim(alpha))) 
		stop("mismatch between dimensions of x and alpha.")

	pd <- vector(length = nrow(x))
	for (i in 1:nrow(x)) pd[i] <- dirichlet1(x[i, ], alpha[i, ])
	pd[apply(x, 1, function(z) any(z < 0 | z > 1))] <- 0
	pd[apply(x, 1, function(z) all.equal(sum(z), 1) != TRUE)] <- 0
	pd
}

dbivnorm <- function(x, y, mu_x, mu_y, sigma_xx, sigma_yy, sigma_xy, log = FALSE) {
  if (length(x) != length(y))
    stop("the x and y vectors must have the same length.")

  if (sigma_xx <= 0 || sigma_yy <= 0) {
    if (log) -Inf else 0
  }
  
  # check positive definiteness
  det_sigma <- sigma_xx*sigma_yy - sigma_xy^2
  if (det_sigma <= 0) {
    if (log) -Inf else 0
  }
  
  # centre observations
  dx <- x - mu_x
  dy <- y - mu_y
  
  # quadratic form: (x - mu)'*Sigma^(-1)*(x - mu)
  # note: for 2x2 matrices, the inverse is: (1/det)*[[d, -b], [-c, a]]
  inv_det <- 1/det_sigma
  quad_form <- inv_det*(sigma_yy*dx^2 - 2*sigma_xy*dx*dy + sigma_xx*dy^2)
  
  # Log density
  log_dens <- -0.5*(log(2*pi) + log(det_sigma) + quad_form)
  
  if (log) log_dens else exp(log_dens)
}

#' Half-t Distribution Density
#'
#' Computes the probability density function of the half-\eqn{t}
#' distribution.
#'
#' @param x A numeric vector of quantiles. Typically non-negative; values
#'   less than zero are allowed but correspond to the density of the underlying
#'   Student-\eqn{t} distribution reflected at zero.
#' @param alpha A positive numeric vector giving the scale parameter.
#' @param nu A positive numeric vector giving the degrees of freedom.
#' @param log Logical; if \code{TRUE}, probabilities are returned on the
#'   log scale.
#'
#' @details
#' The half-Student-\eqn{t} distribution is defined as the distribution of
#' \eqn{|T|}, where \eqn{T} follows a Student-\eqn{t} distribution with
#' \code{nu} degrees of freedom and scale parameter \code{alpha}.
#'
#' The density is
#' \deqn{
#' f(x) = \frac{2}{\mathrm{\alpha}}
#'   \frac{\Gamma((\nu + 1)/2)}{\Gamma(\nu/2)\sqrt{\pi \nu}}
#'   \left(1 + \frac{1}{\nu}\left(\frac{x}{\mathrm{\alpha}}\right)^2\right)^{-(\nu+1)/2},
#' \qquad x \ge 0.
#' }
#'
#' @return
#' A numeric vector giving the density (or log-density) evaluated at \code{x}.
#' The length of the result is the maximum of the lengths of
#' \code{x}, \code{alpha}, and \code{nu}; shorter arguments are recycled.
#'
#' @examples
#' x <- seq(0, 100, length.out = 100)
#' y <- dhalft(x, alpha = 25, nu = 5)
#' plot(x, y, type = "l", main = "Half-t density")
#'
#' # Log-density
#' dhalft(10, alpha = 25, nu = 1, log = TRUE)
#'
#' @seealso
#' \code{\link[stats]{dt}} for the Student-\eqn{t} density.
#'
#' @export
dhalft <- function(x, alpha = 25, nu = 1, log = FALSE) {
  x <- as.vector(x)
  alpha <- as.vector(alpha)
  nu <- as.vector(nu)

  if (any(alpha <= 0)) 
    stop("The alpha (scale) parameter must be positive.")

  NN <- max(length(x), length(alpha), length(nu))
  x <- rep(x, len = NN)
  alpha <- rep(alpha, len = NN)
  nu <- rep(nu, len = NN)

  log_dens <- log(2) - log(alpha) + lgamma((nu + 1)/2) - lgamma(nu/2) - 
    0.5 * log(pi * nu) - (nu + 1)/2 * log(1 + (1/nu) * (x/alpha) * 
    (x/alpha))

  if (log) log_dens else exp(log_dens)
}

pst <- function(q, mu = 0, sigma = 1, nu = 10, lower.tail = TRUE, log.p = FALSE)  {
  q <- as.vector(q)
  mu <- as.vector(mu)
  sigma <- as.vector(sigma)
  nu <- as.vector(nu)
  if (any(sigma <= 0)) 
    stop("The sigma parameter must be positive.")
  if (any(nu <= 0)) 
    stop("The nu parameter must be positive.")

  NN <- max(length(q), length(mu), length(sigma), length(nu))
  q <- rep(q, len = NN)
  mu <- rep(mu, len = NN)
  sigma <- rep(sigma, len = NN)
  nu <- rep(nu, len = NN)
  p <- pt({
      q - mu
  }/sigma, df = nu, lower.tail = lower.tail, log.p = log.p)
  temp <- which(nu > 1e+06)
  p[temp] <- pnorm(q[temp], mu[temp], sigma[temp], lower.tail = lower.tail, 
      log.p = log.p)

  p
}

qst <- function(p, mu = 0, sigma = 1, nu = 10, lower.tail = TRUE, log.p = FALSE) {
  p <- as.vector(p)
  mu <- as.vector(mu)
  sigma <- as.vector(sigma)
  nu <- as.vector(nu)
  if (any(p < 0) || any(p > 1)) 
    stop("p must be in [0,1].")
  if (any(sigma <= 0)) 
    stop("The sigma parameter must be positive.")
  if (any(nu <= 0)) 
    stop("The nu parameter must be positive.")

  NN <- max(length(p), length(mu), length(sigma), length(nu))
  p <- rep(p, len = NN)
  mu <- rep(mu, len = NN)
  sigma <- rep(sigma, len = NN)
  nu <- rep(nu, len = NN)
  q <- mu + sigma * qt(p, df = nu, lower.tail = lower.tail)
  temp <- which(nu > 1e+06)
  q[temp] <- qnorm(p[temp], mu[temp], sigma[temp], lower.tail = lower.tail, log.p = log.p)

  q
}

qtrunc <- function(p, spec, a = -Inf, b = Inf, ...) {
  if (any(p < 0) || any(p > 1)) 
    stop("p must be in [0,1].")
  if (a >= b) 
    stop("Lower bound a is not less than upper bound b.")

  q <- p
  G <- get(paste("p", spec, sep = ""), mode = "function")
  Gin <- get(paste("q", spec, sep = ""), mode = "function")
  q <- Gin(G(a, ...) + p * {G(b, ...) - G(a, ...)}, ...)

  q
}

rtrunc <-function(n, spec, a = -Inf, b = Inf, ...) {
  if (a >= b) 
    stop("Lower bound a is not less than upper bound b.")

  x <- u <- runif(n)
  x <- qtrunc(u, spec, a = a, b = b, ...)

  x
}

#' Random Generation from the Half-t Distribution
#'
#' Generate random deviates from the half-t distribution, which is a 
#' truncated Student's t-distribution restricted to positive values.
#'
#' @param n integer; number of observations to generate.
#' @param alpha numeric; scale parameter. Must be positive. Default is 25.
#'   Vectors are recycled to length \code{n}.
#' @param nu numeric; degrees of freedom parameter. Must be positive. 
#'   Default is 1. Vectors are recycled to length \code{n}.
#'
#' @return A numeric vector of length \code{n} containing random deviates 
#'   from the half-t distribution.
#'
#' @details
#' The half-t distribution is obtained by truncating a Student's t-distribution 
#' with location parameter 0, scale parameter \code{alpha}, and \code{nu} 
#' degrees of freedom to the positive half of the real line.
#'
#' @seealso \code{\link{EXhalft}} for the expected value of the half-t distribution.
#'
#' @examples
#' # Generate 10 random values with default parameters
#' rhalft(10)
#' 
#' # Generate 100 values with custom parameters
#' x <- rhalft(100, alpha = 10, nu = 5)
#' hist(x, main = "Half-t Distribution")
#'
#' @export
rhalft <- function(n, alpha = 25, nu = 1) {
  alpha <- rep(alpha, len = n)
  nu <- rep(nu, len = n)
  if (any(alpha <= 0)) 
    stop("The alpha parameter must be positive.")
  x <- rtrunc(n, "st", a = 0, b = Inf, mu = 0, sigma = alpha, 
    nu = nu)
  
  x
}

#' Expected Value of the Half-t Distribution
#'
#' Calculate the expected value (mean) of the half-t distribution.
#'
#' @param alpha numeric; scale parameter. Must be positive. Default is 25.
#' @param nu numeric; degrees of freedom parameter. Must be greater than 1 
#'   for the expected value to be finite. Default is 1.
#'
#' @return The expected value of the half-t distribution. Returns \code{Inf} 
#'   when \code{nu <= 1}, as the expected value does not exist in this case.
#'
#' @details
#' The expected value of the half-t distribution is given by:
#' \deqn{E[X] = 2\alpha\sqrt{\frac{\nu}{\pi}} \frac{\Gamma((\nu+1)/2)}{\Gamma(\nu/2)(\nu-1)}}
#' for \eqn{\nu > 1}. When \eqn{\nu \leq 1}, the expected value is infinite.
#'
#' @seealso \code{\link{rhalft}} for random generation from the half-t distribution.
#'
#' @examples
#' # Expected value with default parameters (returns Inf since nu = 1)
#' EXhalft()
#' 
#' # Expected value with nu > 1
#' EXhalft(alpha = 10, nu = 5)
#' 
#' # Verify with simulation
#' set.seed(123)
#' mean(rhalft(10000, alpha = 10, nu = 5))
#' EXhalft(alpha = 10, nu = 5)
#'
#' @export
EXhalft <- function(alpha = 25, nu = 1) {
  if (nu <= 1) {
    Inf
  }
  else {
    2*alpha*sqrt(nu/pi)*gamma((nu + 1)/2)/(gamma(nu/2)*(nu - 1))
  }
}
